modmail/config.py

```diff
@@ -1,144 +1,648 @@
     """
-    if USER_CONFIG_PATH:
-        return dict(**toml.load(USER_CONFIG_PATH))
+    if env_prefix is None:
+        env_prefix = ENV_PREFIX
+
+    if env is None:
+        if dotenv_file is not None:
+            env = dotenv.dotenv_values(dotenv_file)
+            env.update(os.environ)
+        else:
+            env = os.environ.copy()
+
+    # get the attributes of the provided class
+    if defaults is None:
+        defaults = defaultdict(lambda: None)
     else:
-        return dict()
-
-
-class BaseSettings(PydanticBaseSettings):
-    class Config:
-        extra = "ignore"
-        env_file = ".env"
-        env_file_encoding = "utf-8"
-
-        @classmethod
-        def customise_sources(
-            cls,
-            init_settings: SettingsSourceCallable,
-            env_settings: SettingsSourceCallable,
-            file_secret_settings: SettingsSourceCallable,
-        ) -> Tuple[SettingsSourceCallable, ...]:
-            return (
-                env_settings,
-                init_settings,
-                file_secret_settings,
-                toml_user_config_source,
-                toml_default_config_source,
+        defaults = defaultdict(lambda: None, defaults.copy())
+
+    kw = defaultdict(lambda: None)  # any missing required vars provide as missing
+
+    for var in attr.fields(klass):
+        if attr.has(var.type):
+            # var is an attrs class too
+            kw[var.name] = _build_class(
+                var.type,
+                env=env,
+                env_prefix=env_prefix + var.name.upper() + "_",
+                defaults=defaults.get(var.name, None),
             )
+        else:
+            kw[var.name] = env.get(env_prefix + var.name.upper(), None)
+            if kw[var.name] is None:
+                if (defa := defaults.get(var.name, None)) is not None and defa is not marshmallow.missing:
+                    kw[var.name] = defaults[var.name]
+                elif var.default is not attr.NOTHING:  # check for var default
+                    kw[var.name] = var.default
+                else:
+                    del kw[var.name]
 
+    return klass(**kw)
 
-class BotConfig(BaseSettings):
-    prefix: str = "?"
-    token: str = None
 
-    class Config:
-        # env_prefix = "bot."
-        allow_mutation = False
+def load_env(env_file: os.PathLike = None, existing_cfg_dict: dict = None) -> dict:
+    """
+    Load a configuration dictionary from the specified env file and environment variables.
+
+    All dependencies for this will always be installed.
+    """
+    if env_file is None:
+        env_file = _CWD / ".env"
+    else:
+        env_file = pathlib.Path(env_file)
 
+    if not existing_cfg_dict:
+        existing_cfg_dict = defaultdict(_generate_default_dict)
 
-class BotMode(BaseSettings):
+    existing_cfg_dict = attr.asdict(_build_class(Cfg, dotenv_file=env_file, defaults=existing_cfg_dict))
+
+    return existing_cfg_dict
```

Comment:
`new_config_dict` for the return?

Reply:
done in ebcaf9485e416619e22f5b12a2e05e4cc0e23ee9